### ISSUE TITLE
New admiralonco helper object: date_source

### ISSUE THREAD

**Issue by rossfarrugia** [pharmaverse/admiralonco#26]
### Feature Idea

date_source object - used to provide the downstream functions with the PD and response dates, as these could come from ADRS parameters, ADEVENT intermediate dataset, or even ADSL for example.

### Relevant Input

```
date_source(
  dataset_name,
  date,
  filter = NULL
)
```

Note: we would only accept DT dates here - reason being as per Step 1 any necessary partial date imputations should already have been applied.

### Relevant Output

See example at https://pharmaverse.github.io/admiral/reference/event_source.html. Output would be an object to define dates as input for our admiralonco derive_param_xxx() functions.

### Reproducible Example/Pseudo Code

Examples:

pd <- date_source(
  dataset_name = “ADSL”,
  date = PDDT,
  filter = NULL
)

resp <- date_source(
  dataset_name = “ADRS”,
  date = ADT,
  filter = PARAMCD == “RSP” & AVALC == “Y”
)

death <- date_source(
  dataset_name = “ADEVENT”,
  date = ADT,
  filter = PARAMCD == “DEATH”
)

**Comment by amsmith214** [comment]
The way I understand these in the `derive_param_tte` function is that they are eventually used by `filter_date_sources`, which expects a `tte_source` object. `filter_date_sources` seems like it would be good to reuse, but will we create an analogue of `filter_date_sources` that expects a `date_source` object?

**Comment by rossfarrugia** [comment]
@amsmith214 do we need filter_date_sources here for ADRS? that sorts multiple dates and takes first or last, but in this case the ADSL var/ADEVENT param/ADRS param would only have one date come through for first response/first PD/death. So whenever date_source gets fed into a downstream parmeter derivation function there would only be one date getting submitted per argument. Or am i missing something?

admittedly i haven't started looking at ADTTE workflow yet (it's on my list) so that might change things later.

**Comment by amsmith214** [comment]
thanks Ross--i think you have the right understanding, and i'm not approaching my issue right. 

for #21 i think `filter_date_sources` would still be useful in establishing the presence of a valid non-PD response assessment ## days from the reference date (e.g., take first and last valid response assessment, and check if at least one of those is ## days from reference), but i think i'm still getting my bearings.

**Comment by rossfarrugia** [comment]
Makes sense @amsmith214 - try it out when you're working on 21 and if we need to update this function from admiral core at all to make it more reusable for this purpose then we could make a new issue there. alternatively you might just be able to use the underlying `filter_extreme` here for your case. I have to say i'm still learning all of the things admiral core has available!

**Comment by clorisxrz** [comment]
@bundfussr Hi Stefan, could you please point me to the file in devel that implements the generation of the object please? Do you mean event_source?

**Comment by bundfussr** [comment]
@clorisxrz , please see https://github.com/pharmaverse/admiral/blob/devel/R/derive_var_extreme_date.R. `date_source()` is defined at the end of the file.

**Comment by rossfarrugia** [comment]
@bundfussr good spot - do you suggest instead then that we move that code out of `derive_var_extreme_date` so that it stands alone a bit more as it has more uses than just for this function? i think it's too hidden away currently in admiral core. if you agree please create a new issue there.

apologies @clorisxrz - i didn't know this existed! can you try out this code and see if it fits our need and if so let me know so i can close this issue and update our ADRS workflow doc accordingly.

**Comment by bundfussr** [comment]
@rossfarrugia , I am not sure if moving `date_source()` in a separate file would help because most users will not go through the list of source files to find a function. However, I agree that it is too hidden. We have this issue also for other functions, e.g., `filter_extreme()`.

**Comment by bundfussr** [comment]
@rossfarrugia , we discussed at the Q&A meeting and decided to restructure the reference page (see https://github.com/pharmaverse/admiral/issues/1105).

I asked Thomas about defining objects like `date_source()` in a separate file. He would prefer defining them in the same file as the function which uses it (i.e., as it is now) or define all objects in one file. I have no strong opinion but would tend to leave it as it is.

**Comment by rossfarrugia** [comment]
ok thanks @bundfussr, that's fine then - sorry slow replying to these this week. should have some more time tomorrow/Friday
